CORE-5145 Merkle Tree Factory Implementation

vlajos opened the pull request:
Merkle Tree Factory implementation and some tests.
API: https://github.com/corda/corda-api/pull/452

vlajos committed:
CORE-5145 Merkle Service

omfR3 commented on libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleProofImpl.kt:
```diff
@@ -5,7 +5,7 @@ import net.corda.v5.crypto.merkle.IndexedMerkleLeaf
 import net.corda.v5.crypto.merkle.MerkleTreeHashDigestProvider
 
-internal class MerkleProofImpl(
+class MerkleProofImpl(
```
~I'm guessing this should implement `MerkleProof` but that's not defined in [this corda-api PR](https://github.com/corda/corda-api/pull/452/files)?~ If there's other PRs this depends on in corda-api, it's probably good to link them in the description (and also link to this PR in the corda-api PR description).

(Update: I see `MerkleProof` is already merged, so of course it wouldn't be defined in that PR :grinning:)

vlajos committed:
CORE-5145 Merkle Service - shorten names

omfR3 committed:
CORE-5145 Add integration test and test.bndrun file. Rewrite some tests to avoid using kotlin.test which causes problems with bnd/OSGi somehow

vlajos committed:
CORE-5145 Merkle Service - rename

omfR3 committed:
CORE-5145: Add missing package-info.java for cipher-suite-impl so it can be used in OSGi integration tests.

Also rename MerkleServiceTest -> MerkleTreeFactoryTest, since the class
under test was renamed.

vlajos committed:
CORE-5145 Merkle: remove createProof from Factory

omfR3 committed:
CORE-5145: Add package-info.java for part of the Merkle tree implementation since it's needed by integration tests. Move integration test to a separate package to solve BND/OSGi problem.

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into vlajos/CORE-5145-merkle-service

mnesbit commented on libs/crypto/merkle/src/main/kotlin/net/corda/crypto/merkle/MerkleTreeFactoryImpl.kt:
```diff
@@ -0,0 +1,54 @@
+class MerkleTreeFactoryImpl @Activate constructor(
+    @Reference(service = DigestService::class)
+    private val digestService: DigestService
+) : MerkleTreeFactory, SingletonSerializeAsToken {
```
Doesn't this need constructor brackets SingletonSerializeAsToken isn't abstract? Has this code been compiled?

omfR3 replied:
Nah that's just an interface now. It compiles and tests ok now.

omfR3 committed:
CORE-5145: MerkleTreeFactoryImpl: use more generic map type and make it non-nullable.

vlajos committed:
CORE-5145: Introduce MerkleTreeHashDigestProviderOption and MerkleTreeHashDigestProviderName

vlajos committed:
CORE-5145: Revert MerkleTreeHashDigestProviderOption and MerkleTreeHashDigestProviderName. Introduce simpler string constants instead.

vlajos committed:
Merge branch 'release/os/5.0' into vlajos/CORE-5145-merkle-service

mnesbit reviewed (dismissed):
OK

vlajos committed:
CORE-5145 Use MerkleTreeHashDigestProviderWithSizeProofSupport

blsemo commented on libs/crypto/cipher-suite-impl/src/main/java/net/corda/cipher/suite/impl/package-info.java:
```diff
@@ -0,0 +1,5 @@
+@Export
```
Should the cipher suite impl really be exported?

omfR3 replied:
Good question. I might have added that file because the integration test depends directly on the impls. Not sure what the best thing to do there is...

vlajos replied:
I have reworked the test so it could be removed.

blsemo commented on libs/crypto/merkle/src/main/java/net/corda/crypto/merkle/package-info.java:
```diff
@@ -0,0 +1,4 @@
+@Export
```
Similar to above - should this be exported?

vlajos replied:
I have reworked the test so it could be removed.

vlajos committed:
CORE-5145: Rework integration test to avoid dependencies on -impl

vlajos committed:
CORE-5145: Use newly introduced constants in the error messages

vlajos committed:
Merge branch 'release/os/5.0' into vlajos/CORE-5145-merkle-service

mcgovc reviewed (approved):
Build changes look fine

blsemo reviewed (approved):
LGTM
We should use it in a flow in a real CorDapp sooner rather than later to make sure that the injection in the flow engine works.
